### PR TITLE
Reshape Tensor

### PR DESCRIPTION
**Pull request by GeorgeNash**
Reshape tensor for double, bool and integer.

Added a reshape to the Boolean Tensor. Didn't want to write my own reshape (edge cases etc...) so I've cast it to an Integer Tensor, done a ND4j reshape and then converted that back to a BooleanTensor.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java [hunk 1]
```diff
@@ -51,6 +51,18 @@ public SimpleBooleanTensor(boolean constant, int[] shape) {
         Arrays.fill(this.data, constant);
     }
 
+    @Override
+    public BooleanTensor reshape(int... newShape) {
+        IntegerTensor convertedForReshape = IntegerTensor.create(asFlatIntegerArray(), shape);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java]: No need to change the data for a reshape. Just need to update the `int[] shape` and `int[] stride`.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     * Returns the supplied vertex with a new shape of the same length
+     */
+    @Override
+    protected DoubleTensor op(DoubleTensor a) {
+        return a.reshape(proposedShape);
+    }
+
+    @Override
+    protected DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
+        Map<Long, DoubleTensor> partialDerivatives = dualNumbers.get(inputVertex).getPartialDerivatives().asMap();
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: This reshaping of the dual numbers should probable happen in the DualNumbers class. There should be a new DualNumber for this vertex and the DualNumber for inputVertex shouldn't be changed or removed.

```
DualNumber reshapedDualNumber = dualNumbers.get(inputVertex).reshape(proposedShape)
dualNumbers.put(getId(), reshapedDualNumber);
```

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
+
+import java.util.Map;
+
+public class ReshapeVertex extends DoubleUnaryOpVertex {
+
+    private int[] proposedShape;
+
+    public ReshapeVertex(DoubleVertex inputVertex, int... proposedShape) {
+        super(inputVertex.getShape(), inputVertex);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: The output shape should be the proposedShape not the same shape as the inputVertex.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        Map<Long, DoubleTensor> partialDerivatives = dualNumbers.get(inputVertex).getPartialDerivatives().asMap();
+        int[] vertexShape = inputVertex.getShape();
+
+        for (Map.Entry<Long, DoubleTensor> partialDerivative : partialDerivatives.entrySet()) {
+            int[] shape = partialDerivative.getValue().getShape();
+            int[] wrtShape = extractWrtShape(shape, vertexShape);
+            int[] newPartialShape = TensorShape.concat(proposedShape, wrtShape);
+
+            DoubleTensor reshapedPartialDerivative = partialDerivative.getValue().reshape(newPartialShape);
+            partialDerivative.setValue(reshapedPartialDerivative);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: Make sure you don't mutate the partialDerivative of the parent. We should look into making the PartialDerivative class immutable.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerReshapeVertex.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.intgr.IntegerVertex;
+
+public class IntegerReshapeVertex extends IntegerUnaryOpVertex {
+
+    private int[] proposedShape;
+
+    public IntegerReshapeVertex(IntegerVertex inputVertex, int... proposedShape) {
+        super(inputVertex.getShape(), inputVertex);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerReshapeVertex.java]: Output shape is the proposed shape, not the inputVertexShape.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/BooleanTensor.java [hunk 6]
```diff
@@ -26,6 +26,8 @@ static BooleanTensor placeHolder(int[] shape) {
         return new SimpleBooleanTensor(shape);
     }
 
+    BooleanTensor reshape(int... newShape);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/bool/BooleanTensor.java]: Can we make reshape part of the Tensor interface?

` Tensor<T> reshape(int... newShape);`

File: keanu-project/src/test/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensorTest.java [hunk 7]
```diff
@@ -151,4 +151,11 @@ public void canConvertToIntegers() {
         assertArrayEquals(new int[]{1, 0, 1, 0}, matrixA.asFlatIntegerArray());
     }
 
+    @Test
+    public void canReshape() {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensorTest.java]: Nice! Not sure how Boolean didn't have a reshape...

File: keanu-project/src/main/java/io/improbable/keanu/vertices/bool/nonprobabilistic/operators/unary/BoolReshapeVertex.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import io.improbable.keanu.tensor.bool.BooleanTensor;
+import io.improbable.keanu.vertices.bool.BoolVertex;
+
+public class BoolReshapeVertex extends BoolUnaryOpVertex<BooleanTensor> {
+
+    private int[] proposedShape;
+
+    public BoolReshapeVertex(BoolVertex inputVertex, int... proposedShape) {
+        super(inputVertex.getShape(), inputVertex);
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/bool/nonprobabilistic/operators/unary/BoolReshapeVertex.java]: Nice spot, changed to `proposedShape`

File: keanu-project/src/test/java/io/improbable/keanu/vertices/bool/BoolVertexTest.java [hunk 9]
```diff
@@ -7,12 +7,15 @@
 import io.improbable.keanu.vertices.Vertex;
 import io.improbable.keanu.vertices.bool.nonprobabilistic.CastBoolVertex;
 import io.improbable.keanu.vertices.bool.nonprobabilistic.ConstantBoolVertex;
+import io.improbable.keanu.vertices.bool.nonprobabilistic.operators.unary.BoolReshapeVertex;
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/bool/BoolVertexTest.java]: removed

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java (hunk 1 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java]: In `ND4JDoubleTensor` the reshape creates a dup and doesn't affect the original, so I've gone for: 

```
    @Override
    public BooleanTensor reshape(int... newShape) {
        if (TensorShape.getLength(shape) != TensorShape.getLength(newShape)) {
            throw new IllegalArgumentException("Cannot reshape a tensor to a shape of different length");
        }
        return new SimpleBooleanTensor(data, newShape);
    }
```

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java (hunk 2 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: Nice, looks much cleaner. changed.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java (hunk 3 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: Good spot, changed

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java (hunk 4 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: Oops - I was doing a ReshapeInPlaceVertex by mistake...

Changed to not mutate the parent

File: keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerReshapeVertex.java (hunk 5 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerReshapeVertex.java]: nice, changed

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/BooleanTensor.java (hunk 6 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/bool/BooleanTensor.java]: Yep, added

File: keanu-project/src/test/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensorTest.java (hunk 7 above)
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensorTest.java]: poor guy was feeling left out

**Pull request merged**